stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ChannelControllerImpl.kt

```diff
@@ -1277,12 +1278,33 @@ internal class ChannelControllerImpl(
     fun upsertMember(member: Member) = upsertMembers(listOf(member))
 
     private fun updateReads(reads: List<ChannelUserRead>) {
+        val currentUserId = domainImpl.currentUser.id
         val previousUserIdToReadMap = _reads.value
         val incomingUserIdToReadMap = reads.associateBy(ChannelUserRead::getUserId).toMutableMap()
 
-        incomingUserIdToReadMap[domainImpl.currentUser.id]?.let {
-            _read.value = it
-            _unreadCount.value = it.unreadMessages
+        /**
+         * It's possible that the data coming back from the online channel query has a last read date that's
+         * before what we've last pushed to the UI. We want to ignore this, as it will cause an unread state
+         * to show in the channel list.
+         */
+        val incomingUserRead = incomingUserIdToReadMap[domainImpl.currentUser.id] ?: return
+
+        // the previous last Read date that is most current
+        val previousLastRead = _read.value?.lastRead ?: previousUserIdToReadMap[currentUserId]?.lastRead
+
+        // Use AFTER to determine if the incoming read is more current.
+        // This prevents updates if it's BEFORE or EQUAL TO the previous Read.
+        val shouldUpdateByIncoming = previousLastRead == null || incomingUserRead.lastRead?.inOffsetWith(
+            previousLastRead,
+            OFFSET_EVENT_TIME
+        ) == true
+
+        if (shouldUpdateByIncoming) {
+            _read.value = incomingUserRead
+            _unreadCount.value = incomingUserRead.unreadMessages
+        } else {
+            // if the previous Read was more current, replace the item in the update map
+            incomingUserIdToReadMap[currentUserId] = ChannelUserRead(domainImpl.currentUser, previousLastRead)
```

Comment:
Why do we need to update this map here? Doesn't it already contain the `previousLastRead`?

Reply:
Just to be sure that we have always fresh correct data